### PR TITLE
P20-1273: Track user level progress with current locale

### PR DESCRIPTION
**Pull request by artem-vavilov**
The RED team often wants to know which language a user selects after completing a level.
Currently, we only have the locale field in the users table, which records the user's first language selection site-wide. However, we need to track language preferences at the user-level granularity as different curricula are translated.

## Links
- JIRA task: [P20-1273](https://codedotorg.atlassian.net/browse/P20-1273)
- Migration PR: https://github.com/code-dot-org/code-dot-org/pull/62302

### CODE REVIEW

**Review by allison-code-dot-org — APPROVED** [review]

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -2080,7 +2080,18 @@ def self.track_proficiency(user_id, script_id, level_id)
 
   # The synchronous handler for the track_level_progress helper.
   # @return [UserLevel]
-  def self.track_level_progress(user_id:, level_id:, script_id:, new_result:, submitted:, level_source_id:, pairing_user_ids: nil, is_navigator: false, time_spent: nil)
+  def self.track_level_progress(
```
Comment (mgc1194) [inline dashboard/app/models/user.rb:2083]: Should we add a test case for the locale in the user model test?

**Review by mgc1194 — APPROVED** [review]

File: dashboard/app/models/user.rb (hunk 1 above)
Comment (artem-vavilov) [reply dashboard/app/models/user.rb:2083]: We probably should, but we don't really need to. I'll add one.

File: dashboard/test/test_helper.rb [hunk 2]
```diff
@@ -261,6 +261,10 @@ def assert_authentication_option(actual_option, expected_values)
     end
   end
 
+  def set_request_locale(locale)
```
Comment (mgc1194) [inline dashboard/test/test_helper.rb:264]: I thought we already had a way to mock the locale, but it looks like we were just forcing the request value to a certain locale:
https://github.com/code-dot-org/code-dot-org/blob/d16ce7f8772b6666c0914cf51654f4b1f5264cc3/dashboard/test/controllers/home_controller_test.rb#L185

**Review by mgc1194 — APPROVED** [review]

**Comment by cat5inthecradle** [comment]
Do we ever use this object to define or change behavior on the site, or is this purely analytical?

Wondering if there's risk of us trying to use 'locale' inappropriately in the future.

**Comment by artem-vavilov** [comment]
> Do we ever use this object to define or change behavior on the site, or is this purely analytical?
> 
> Wondering if there's risk of us trying to use 'locale' inappropriately in the future.

This is purely analytical

**Pull request closed without merging**